### PR TITLE
aarch64 and x86_64 alpine APK client build

### PR DESCRIPTION
**Pull request by lbuga-dn**

### CODE REVIEW

File: apk_build/APKBUILD [hunk 1]
```diff
@@ -0,0 +1,81 @@
+pkgname=stronglahav
+pkgver=1.0.0
+pkgrel=1
+pkgdesc="StrongLahav"
+url="https://www.strongswan.org/"
+arch="aarch64"
+pkgusers="ipsec"
+pkggroups="ipsec"
+license="GPL-2.0-or-later WITH OpenSSL-Exception"
+options="!check" # failing tests
```
Comment (lschlesinger-dn) [inline apk_build/APKBUILD:10]: Why the tests are failing?

File: containers/Makefile [hunk 2]
```diff
@@ -35,9 +35,13 @@ grpc_clone:
 build-builder: grpc_clone
 	docker build -f docker/server/Dockerfile --tag ${BUILDER_IMAGE_NAME} .
 
-build-builder-client:
+build-builder-client: .BIN_FMT
```
Comment (lschlesinger-dn) [inline containers/Makefile:38]: Why the builder needs multiarch support?

File: containers/Makefile [hunk 3]
```diff
@@ -71,12 +75,13 @@ clean-strongswan:
 # 1. Configs StrongSwan for client
 # 2. Builds StrongSwan from source
 # 3. Builds a debian containing StrongSwan
-build-strongswan-client:
+build-strongswan-client: .BIN_FMT
 	docker run --rm \
 		-v ${strongswan_dir}/:/strongswan \
+		-v ${strongswan_dir}/../packages:/root/packages/strongswan/aarch64 \
```
Comment (lschlesinger-dn) [inline containers/Makefile:81]: Leftovers?

File: containers/Makefile [hunk 4]
```diff
@@ -71,12 +75,13 @@ clean-strongswan:
...
 		-e CLIENT_ARCH=${CLIENT_ARCH} \
 		${CLIENT_BUILDER_IMAGE_NAME} \
-		bash -c '/strongswan/containers/scripts/build_strongswan.sh build-strongswan-client'
+		bash -c '/strongswan/containers/scripts/build_strongswan.sh build-strongswan-client-apk'
```
Comment (lschlesinger-dn) [inline containers/Makefile:84]: If `build-strongswan-client` is no longer used then remove it from `build_strongswan.sh`

File: src/libstrongswan/utils/backtrace.c [hunk 5]
```diff
@@ -408,6 +408,8 @@ static void print_sourceline(FILE *file, char *filename, void *ptr, void *base)
 void backtrace_init() {}
 void backtrace_deinit() {}
 
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wunused-function"
```
Comment (lschlesinger-dn) [inline src/libstrongswan/utils/backtrace.c:412]: What is missing from the environment that causes this function to not be used?

File: containers/Makefile [hunk 6]
```diff
@@ -71,12 +75,13 @@ clean-strongswan:
 # 1. Configs StrongSwan for client
 # 2. Builds StrongSwan from source
 # 3. Builds a debian containing StrongSwan
```
Comment (lschlesinger-dn) [inline containers/Makefile:77]: Update comment :)

File: containers/scripts/build_strongswan.sh [hunk 7]
```diff
@@ -274,6 +278,13 @@ case "x${BUILD_MODE}" in
 		build_strongswan_deb_for_client
 		;;
 
+	"xbuild-strongswan-client-apk")
```
Comment (lschlesinger-dn) [inline containers/scripts/build_strongswan.sh:281]: Also update 'clean_strongswan' to clean the .apk, the apk_build directory and any other file created during the build

File: containers/scripts/build_strongswan.sh [hunk 8]
```diff
@@ -274,6 +278,13 @@ case "x${BUILD_MODE}" in
...
+		cd /strongswan
+		write_stronglahav_version
+		cd apk_build && abuild -F
+		cd -
```
Comment (lschlesinger-dn) [inline containers/scripts/build_strongswan.sh:285]: Please move these lines to a new function

File: configure.ac [hunk 9]
```diff
@@ -371,7 +371,7 @@ AC_SUBST(PLUGIN_CFLAGS)
 AC_PROG_CC
 AM_PROG_CC_C_O
 
-AC_LIB_PREFIX
+#AC_LIB_PREFIX
```
Comment (lschlesinger-dn) [inline configure.ac:374]: Why commented this out? What is the impact of not having it?
Also note that from https://wiki.strongswan.org/issues/652 installing some extra packages should solve issues with this macro

Comment (lbuga-dn) [reply configure.ac:374]: `error: possibly undefined macro: AC_LIB_PREFIX`

File: apk_build/APKBUILD (hunk 1 above)
Comment (lbuga-dn) [reply apk_build/APKBUILD:10]: Feel free to check and fix, I won't mind:

```Passed all 2 'prf_plus' test cases
  Running suite 'http fetcher':
    Running case 'GET': +++-
      Failure in 'test_get': service != NULL (suites/test_fetch_http.c:211, i = 3)
    Running case 'POST': ++-
      Failure in 'test_post': service != NULL (suites/test_fetch_http.c:250, i = 2)
    Running case 'response code': +++++
  Passed 1/3 'http fetcher' test cases
  Running suite 'ed25519':
    Running case 'ed25519_sign': +++++
    Running case 'ed25519_gen': +
    Running case 'ed25519_fail': +
    Running case 'ed25519_speed': +
  Passed all 4 'ed25519' test cases
  Running suite 'ed448':
    Running case 'ed448_sign': ++++++++
    Running case 'ed448_gen': +
    Running case 'ed448_fail': +
    Running case 'ed448_speed': +
  Passed all 4 'ed448' test cases
  Running suite 'signature params':
    Running case 'rsa/pss parse': +++++++++++++++
    Running case 'rsa/pss build': +++++++++++++
    Running case 'rsa/pss salt len': ++++++++++
    Running case 'params compare': +++++++++
    Running case 'params clone': ++++++
    Running case 'params clear': ++++++
    Running case 'parse': +++++++++++++++++++
    Running case 'build': ++++++++++++++++
  Passed all 8 'signature params' test cases
  Running suite 'metadata':
    Running case 'integer types': ++++++
  Passed all 1 'metadata' test cases
  Running suite 'metadata_set':
    Running case 'create/destroy': ++
    Running case 'put/get': +++++++++
    Running case 'clone': +++
  Passed all 3 'metadata_set' test cases
Passed 46 of 49 'libstrongswan' suites
FAIL: libstrongswan_tests
==================
1 of 1 test failed
==================
```

File: src/libstrongswan/utils/backtrace.c (hunk 5 above)
Comment (lbuga-dn) [reply src/libstrongswan/utils/backtrace.c:412]: Obviously some `#ifdef` mess in there.

Not enough time to check / fix all StrongSwan issues.
Please keep comments relevant to PR scope which was to build for `aarch64`.

File: containers/Makefile (hunk 3 above)
Comment (lbuga-dn) [reply containers/Makefile:81]: This is where you get the `xxx.apk` and other artifacts of `abuild`.

File: src/libstrongswan/utils/backtrace.c (hunk 5 above)
Comment (lschlesinger-dn) [reply src/libstrongswan/utils/backtrace.c:412]: This is very much relevant to this PR... If the apk you're shipping doesn't have support for printing backtraces, then we need to know about it.

File: apk_build/APKBUILD (hunk 1 above)
Comment (lschlesinger-dn) [reply apk_build/APKBUILD:10]: This is your PR.....

File: containers/Makefile [hunk 10]
```diff
@@ -35,9 +35,13 @@ grpc_clone:
 build-builder: grpc_clone
 	docker build -f docker/server/Dockerfile --tag ${BUILDER_IMAGE_NAME} .
 
-build-builder-client:
+build-builder-client: .BIN_FMT
 	docker build -f docker/client/Dockerfile --tag ${CLIENT_BUILDER_IMAGE_NAME} .
 
+.BIN_FMT:
+	docker run --rm --privileged multiarch/qemu-user-static --reset -p yes
+	touch .BIN_FMT
```
Comment (lschlesinger-dn) [inline containers/Makefile:43]: Restarting the host will keep the `.BIN_FMT` file, but the binfmt entries will be removed.
In my machine it takes less than a second to run this docker command, I don't see a reason to enforce it to run only once.

File: containers/Makefile (hunk 3 above)
Comment (lschlesinger-dn) [reply containers/Makefile:81]: I see. In this case please don't place it outside this repo's directory, no need to pollute the host's FS if can avoid it.
Best just copy the whole directory into `/strongswan/containers/`, the same place where the ubuntu debian is saved to.

File: src/libstrongswan/utils/backtrace.c (hunk 5 above)
Comment (lbuga-dn) [reply src/libstrongswan/utils/backtrace.c:412]: most likely `musl` libc implementation does not provide support for this, thus we won't have backtrace support.

File: apk_build/APKBUILD (hunk 1 above)
Comment (lbuga-dn) [reply apk_build/APKBUILD:10]: ... where I do not have functional code changes. I'm just doing the build.

Comment (lbuga-dn) [reply apk_build/APKBUILD:10]: same test failures in `main` branch with `86_64` build, so I rest my case on this

**Review by lschlesinger-dn — APPROVED** [review]

**Pull request merged**